### PR TITLE
Beta release fixes and feature.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/dart-lang/code_builder/issues/145.
Closes https://github.com/dart-lang/code_builder/issues/146.
Closes https://github.com/dart-lang/code_builder/issues/147.

### CODE REVIEW

**Comment by matanlurey** [comment]
Ping @yjbanov plz 🗡

File: lib/src/base.dart [hunk 1]
```diff
@@ -7,3 +7,17 @@ import 'visitors.dart';
...
+Spec lazySpec(Spec Function() generate) => new _LazySpec(generate);
+
+class _LazySpec implements Spec {
+  final Spec Function() generate;
+
+  const _LazySpec(this.generate);
+
+  @override
+  R accept<R>(SpecVisitor<R> visitor, [R context]) {
+    return generate().accept(visitor, context);
```
Comment (yjbanov) [inline lib/src/base.dart]: This not only generates lazily, but also generates _every time_ it's visited. Is it intentional?

File: lib/src/specs/code.dart [hunk 2]
```diff
@@ -111,7 +115,33 @@ abstract class CodeEmitter implements CodeVisitor<StringSink> {
...
-/// Represents a simple, literal [code] block to be inserted as-is.
+/// Represents a code block that requires lazy visiting.
+class LazyCode implements Code {
+  final Spec Function(SpecVisitor) generate;
+
+  const LazyCode._(this.generate);
+
+  @override
+  R accept<R>(CodeVisitor<R> visitor, [R context]) {
+    return generate(visitor).accept(visitor, context);
```
Comment (yjbanov) [inline lib/src/specs/code.dart]: same here

File: lib/src/specs/code.dart [hunk 3]
```diff
@@ -111,7 +115,33 @@ abstract class CodeEmitter implements CodeVisitor<StringSink> {
...
+Code lazyCode(Code Function() generate) => new _LazyCode(generate);
+
+class _LazyCode implements Code {
+  final Code Function() generate;
+
+  const _LazyCode(this.generate);
+
+  @override
+  R accept<R>(CodeVisitor<R> visitor, [R context]) {
+    return generate().accept(visitor, context);
```
Comment (yjbanov) [inline lib/src/specs/code.dart]: ditto

File: lib/src/base.dart (hunk 1 above)
Comment (matanlurey) [reply lib/src/base.dart]: Yeah. We could cache it, but we wouldn't know if the function is stateless or not.

File: lib/src/specs/code.dart (hunk 2 above)
Comment (matanlurey) [reply lib/src/specs/code.dart]: Ditto.

File: lib/src/specs/code.dart (hunk 3 above)
Comment (matanlurey) [reply lib/src/specs/code.dart]: Ditto.

**Pull request merged**